File: programmers/python/level3/dp/way_to_school.py
## https://programmers.co.kr/learn/courses/30/lessons/42898
def solution(m, n, puddles):
    graph = [[-1] * m for _ in range(n)]
    for puddle in puddles:
        x, y = puddle
        graph[y-1][x-1] = 0

    graph[0][0] = 1
    for x in range(1, m):
        if graph[0][x] == 0 :
            for y in range(x+1, m):
                graph[0][y] = 0
            break
        else:
            graph[0][x] = 1
    
    for x in range(1, n):
        if graph[x][0] == 0 :
            for y in range(x+1, n):
                graph[y][0] = 0
            break
        else:
            graph[x][0] = 1


    for i in range(n):
        for j in range(m):
            if graph[i][j] != -1:
                continue
            graph[i][j] = (graph[i][j-1] + graph[i-1][j]) % 1000000007

    return graph[n-1][m-1]

print(solution(4,3,[[2,2]]))

print(1000000008 % 1000000007)

File: programmers/java/src/level3/dp/WayToSchool.java
package level3.dp;

//https://programmers.co.kr/learn/courses/30/lessons/42898
public class WayToSchool {
    public static void main(String[] args) {
        WayToSchool wts = new WayToSchool();
        System.out.println(wts.solution(4, 3, new int[][]{{2,2}}));
    }

    public int solution(int m, int n, int[][] puddles) {
        int[][] graph = new int[n+1][m+1];
        for(int[] p : puddles) {
            graph[p[1]][p[0]] = -1;
        }

        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <=m; j++) {
                if(i==1 && j==1) {
                    graph[i][j] = 1;
                    continue;
                }
                if(graph[i][j] == -1) {
                    graph[i][j] = 0;
                    continue;
                }
                graph[i][j] = (graph[i-1][j] + graph[i][j-1]) % 1000000007;
            }
        }

        return graph[n][m];
    }
}
